### PR TITLE
Feature/40013

### PR DESCRIPTION
**Pull request by Sumitsuccessivetech**
- Ticket No. [#40013](https://redmine.successive.tech/issues/40013)
- SetUp all The Entities

### CODE REVIEW

File: src/controllers/user/controller.ts [hunk 1]
```diff
@@ -14,112 +18,141 @@ class UserController {
...
-                        message: 'Password not Mtached with DB'
-                    })
+
+                const { password } = userData;
+
+                if (password !== req.body.password) {
+                    res.status(401).send({
+                        err: 'Invalid Password',
+                        code: 401
+                    });
```
Comment (chiragSuccessive) [inline src/controllers/user/controller.ts]: Please use your errorHandler for sending error in response.

File: src/controllers/user/controller.ts [hunk 2]
```diff
@@ -14,112 +18,141 @@ class UserController {
...
-                    status: 404,
-                    message: 'Email Not Exist in DB'
-                })
-            };
-        });
-    }
-    catch ( err ) {
-        res.send( err );
+
+                const token = jwt.sign(userData.toJSON(),'qwertyuiopasdfghjklzxcvbnm123456');
```
Comment (chiragSuccessive) [inline src/controllers/user/controller.ts]: toJSON() ? why are we using it here?

File: src/controllers/user/controller.ts [hunk 3]
```diff
@@ -14,112 +18,141 @@ class UserController {
...
+                res.send({
+                    message: 'Login Successfull',
+                    status: 200,
+                    'token': token
```
Comment (chiragSuccessive) [inline src/controllers/user/controller.ts]: `'token'` could be written as `token`

File: src/controllers/user/controller.ts [hunk 4]
```diff
@@ -14,112 +18,141 @@ class UserController {
...
+
+   public async me(req: IRequest, res: Response, next: NextFunction) {
+      const id = req.query;
+        const user = new UserRepository();
+
+        await user.get({ id })
+            .then((data) => {
+                res.status(200).send({
+                    message: 'User Fetched successfully',
+                    'data': { data },
```
Comment (chiragSuccessive) [inline src/controllers/user/controller.ts]: Remove string from _**data**_ key.
`'data': { data }` why we are sending **_data_** in object, when _**data**_ itself is an object.

File: src/controllers/user/controller.ts [hunk 5]
```diff
@@ -14,112 +18,141 @@ class UserController {
...
+                    status: 404
 
-                },
-                {
-                    name: 'user2',
-                }]
+                })
+            }
+            res.send({
+                message: 'trainee created successfully',
```
Comment (chiragSuccessive) [inline src/controllers/user/controller.ts]: Its a user controller, so message should be 'user created successfully'

File: src/controllers/user/controller.ts [hunk 6]
```diff
@@ -14,112 +18,141 @@ class UserController {
...
-            console.log('Inside put method of Trainee');
+            const id = req.params.id;
+            const user = await this.userRepository.userUpdate(id, req.headers.user);
+            if (!id) {
+                next({
+                    message: 'User Not Updated',
+                    error: 'id is Required',
+                    status: 404
+                })
+            }
```
Comment (chiragSuccessive) [inline src/controllers/user/controller.ts]: id should be check before the operation of update applied.

File: src/controllers/user/controller.ts [hunk 7]
```diff
@@ -14,112 +20,137 @@ class UserController {
...
-    catch ( err ) {
-        res.send( err );
+                const expDate = new Date();
+                const payLoad = {
+                    name: 'sumit',
+                    iss: new Date(),
+                    exp: expDate.setDate(expDate.getDate() + 7),
+                    email: "sumit.upadhyay@successive.tech",
+                    role: 'trainee'
+                }
```
Comment (chiragSuccessive) [inline src/controllers/user/controller.ts]: Use dynamic payload.

File: src/controllers/user/controller.ts [hunk 8]
```diff
@@ -14,112 +20,137 @@ class UserController {
...
-                {
-                    name: 'user2',
-                }]
-            });
+            const id = req.params.id;
+            if (!id) {
+                next({
+                    message: 'id is Required',
+                    error: 404,
+                })
```
Comment (chiragSuccessive) [inline src/controllers/user/controller.ts]: Unnecessary checks, as this has been already checked in the validation middleware.

File: src/controllers/user/controller.ts [hunk 9]
```diff
@@ -14,112 +20,137 @@ class UserController {
...
```
Comment (chiragSuccessive) [inline src/controllers/user/controller.ts]: Unnecessary checks, as this has been already checked in the validation middleware.

File: src/repositories/versionable/VersionableRepository.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        console.log('previous: ', previous);
+        if (previous) {
+            await this.invalidate(data.id);
+        } else {
+            return undefined;
+        }
+        const newData = Object.assign(JSON.parse(JSON.stringify(previous)), data);
+        newData._id = VersionableRepository.generateObjectId();
+        delete newData.deletedAt;
+        newData.updatedAt;
```
Comment (chiragSuccessive) [inline src/repositories/versionable/VersionableRepository.ts]: Add date in **_updatedAt_** field.

**Review by chiragSuccessive — CHANGES_REQUESTED** [review]

File: src/controllers/user/controller.ts (hunk 7 above)
Comment (Sumitsuccessivetech) [reply src/controllers/user/controller.ts]: I am not getting How to add a dynamic payload. I tried but I did not get it. Please help me to solve it.

**Review by chiragSuccessive — APPROVED** [review]

**Pull request merged**